File: services/api/src/api/models/turn_models.py
"""Turn-related Pydantic models for request/response validation."""

from __future__ import annotations

import re

from pydantic import BaseModel, Field
from pydantic import field_validator

from src.api.models.envelope import ApiEnvelope


def _word_count(text: str) -> int:
    return len(re.findall(r"\S+", text.strip()))


class CoachingDimension(BaseModel):
    """Single rubric dimension in coaching feedback."""

    label: str = Field(..., description="Rubric dimension label")
    score: int = Field(..., ge=1, le=5, description="Score from 1 to 5")
    tip: str = Field(..., description="Actionable short tip (<= 25 words)")

    @field_validator("tip")
    @classmethod
    def validate_tip_length(cls, value: str) -> str:
        if _word_count(value) > 25:
            raise ValueError("tip must be 25 words or fewer")
        return value


class CoachingFeedback(BaseModel):
    """Structured coaching feedback aligned to rubric dimensions."""

    dimensions: list[CoachingDimension] = Field(
        ..., description="List of rubric dimension scores and tips"
    )
    summary_tip: str = Field(..., description="One-sentence summary tip (<= 30 words)")

    @field_validator("summary_tip")
    @classmethod
    def validate_summary_tip_length(cls, value: str) -> str:
        if _word_count(value) > 30:
            raise ValueError("summary_tip must be 30 words or fewer")
        return value


class TurnResponseData(BaseModel):
    """Response data for a turn submission.

    Represents the result of processing a user's audio answer through
    the turn pipeline (STT → LLM → TTS).
    """

    transcript: str = Field(
        ...,
        description="Speech-to-text transcription of the user's answer",
        examples=[
            "I would approach this problem by breaking it down into smaller parts."
        ],
    )

    assistant_text: str | None = Field(
        default=None,
        description="LLM-generated follow-up question or response",
        examples=["That's a good start. Can you elaborate on your approach?"],
    )

    tts_audio_url: str | None = Field(
        default=None,
        description="URL to fetch TTS audio of assistant_text",
        examples=["/tts/550e8400-e29b-41d4-a716-446655440000"],
    )

    coaching_feedback: CoachingFeedback | None = Field(
        default=None,
        description="Structured per-turn coaching feedback aligned to rubric",
    )

    timings: dict[str, float] = Field(
        ...,
        description="Pipeline stage timings in milliseconds",
        examples=[
            {"upload_ms": 120.5, "stt_ms": 820.3, "llm_ms": 1200.0, "total_ms": 2140.8}
        ],
    )

    is_complete: bool = Field(
        default=False,
        description="Whether this was the final turn (session complete)",
        examples=[False, True],
    )

    question_number: int = Field(
        ...,
        description="Current question number (1-indexed)",
        examples=[1, 2, 3, 5],
    )

    total_questions: int = Field(
        ...,
        description="Total configured questions for the session",
        examples=[3, 5, 10],
    )


# Type alias for the complete turn response with envelope
TurnResponse = ApiEnvelope[TurnResponseData]

File: apps/mobile/lib/core/models/turn_models.dart
import 'package:json_annotation/json_annotation.dart';

part 'turn_models.g.dart';

/// Coaching rubric dimension score and tip.
@JsonSerializable()
class CoachingDimension {
  /// Creates a [CoachingDimension].
  const CoachingDimension({
    required this.label,
    required this.score,
    required this.tip,
  });

  /// Creates a [CoachingDimension] from JSON.
  factory CoachingDimension.fromJson(Map<String, dynamic> json) =>
      _$CoachingDimensionFromJson(json);

  /// Rubric dimension label (e.g., Clarity).
  final String label;

  /// Score from 1 to 5.
  final int score;

  /// Actionable short tip for this dimension.
  final String tip;

  /// Converts this [CoachingDimension] to JSON.
  Map<String, dynamic> toJson() => _$CoachingDimensionToJson(this);
}

/// Structured coaching feedback returned per turn.
@JsonSerializable()
class CoachingFeedback {
  /// Creates a [CoachingFeedback].
  const CoachingFeedback({
    required this.dimensions,
    required this.summaryTip,
  });

  /// Creates a [CoachingFeedback] from JSON.
  factory CoachingFeedback.fromJson(Map<String, dynamic> json) =>
      _$CoachingFeedbackFromJson(json);

  /// Rubric dimensions with score and tip.
  final List<CoachingDimension> dimensions;

  /// Most impactful one-line improvement tip.
  @JsonKey(name: 'summary_tip')
  final String summaryTip;

  /// Converts this [CoachingFeedback] to JSON.
  Map<String, dynamic> toJson() => _$CoachingFeedbackToJson(this);
}

/// Structured end-of-session summary returned on final turn.
@JsonSerializable()
class SessionSummary {
  /// Creates a [SessionSummary].
  const SessionSummary({
    required this.overallAssessment,
    required this.strengths,
    required this.improvements,
    required this.averageScores,
    this.recommendedActions = const <String>[],
  });

  /// Creates a [SessionSummary] from JSON.
  factory SessionSummary.fromJson(Map<String, dynamic> json) =>
      _$SessionSummaryFromJson(json);

  /// 2-3 sentence overall review.
  @JsonKey(name: 'overall_assessment')
  final String overallAssessment;

  /// Concrete strengths (1-3 items).
  final List<String> strengths;

  /// Growth-oriented improvements (1-3 items).
  final List<String> improvements;

  /// Per-dimension average scores.
  @JsonKey(name: 'average_scores')
  final Map<String, double> averageScores;

  /// Concrete next steps tied to rubric weaknesses (2-4 items when present).
  @JsonKey(name: 'recommended_actions', defaultValue: <String>[])
  final List<String> recommendedActions;

  /// Converts this [SessionSummary] to JSON.
  Map<String, dynamic> toJson() => _$SessionSummaryToJson(this);
}

/// Response data from POST /turn containing transcript and metadata.
@JsonSerializable()
class TurnResponseData {
  /// Creates a [TurnResponseData].
  const TurnResponseData({
    required this.transcript,
    required this.timings,
    required this.questionNumber,
    required this.totalQuestions,
    this.assistantText,
    this.ttsAudioUrl,
    this.coachingFeedback,
    this.sessionSummary,
    this.isComplete = false,
  });

  /// Creates a [TurnResponseData] from JSON.
  factory TurnResponseData.fromJson(Map<String, dynamic> json) =>
      _$TurnResponseDataFromJson(json);

  /// The transcribed text from the user's audio.
  final String transcript;

  /// The assistant's text response (LLM-generated follow-up question).
  @JsonKey(name: 'assistant_text')
  final String? assistantText;

  /// URL to the TTS audio file (null until TTS is integrated).
  @JsonKey(name: 'tts_audio_url')
  final String? ttsAudioUrl;

  /// Structured coaching feedback for this turn.
  @JsonKey(name: 'coaching_feedback')
  final CoachingFeedback? coachingFeedback;

  /// End-of-session summary (final turn only).
  @JsonKey(name: 'session_summary')
  final SessionSummary? sessionSummary;

  /// Timing information for each processing stage in milliseconds.
  final Map<String, double> timings;

  /// Whether this was the final turn (session complete).
  @JsonKey(name: 'is_complete')
  final bool isComplete;

  /// Current question number (1-indexed).
  @JsonKey(name: 'question_number')
  final int questionNumber;

  /// Total configured questions for the session.
  @JsonKey(name: 'total_questions')
  final int totalQuestions;

  /// Converts this [TurnResponseData] to JSON.
  Map<String, dynamic> toJson() => _$TurnResponseDataToJson(this);
}
